### ISSUE TITLE
Cut cross entropy for pretraining and finetuning

### ISSUE THREAD

**Issue by Aceticia** [Lightning-AI/litgpt#2163]
[Cut cross entropy](https://arxiv.org/pdf/2411.09009) has been shown in my own work to be really easy to use and saves a lot of memory when training models with large vocab sizes. Therefore I think adding it to litGPT would help people fit bigger models more effortlessly. The main idea is that with some clever intermediate states we can avoid materializing the full logit tensor (Batch x sequence x vocab size), thus saving a lot of memory. The loss is instead computed using the final hidden states and the weight matrix.

As to how this can be integrated, I think here's how it can happen:
1. The GPT model needs an option to return not the logits but the output projection weight and the final hiddens.
2. An option can be added to use for the cross entropy for pretrain and finetune script.

I'm happy to help with implementing this if folks think this is a worthy addition to litGPT.
